feat(ndt_scan_matcher): implement tpe sampler in ndt_align_srv

SakodaShintaro opened the pull request:
## Description

The ndt service for initial position estimation has been changed from random search to search using TPE.

## Related links
https://github.com/autowarefoundation/autoware_launch/pull/602

https://github.com/tier4/autoware-spell-check-dict/pull/623

## Summary of Changes

![tpe](https://github.com/autowarefoundation/autoware.universe/assets/28504069/357b2829-5f5e-4775-9c82-e53bee8bb8aa)

## Tests performed

Tested by 

1. logging simulator about the [sample rosbag](https://autowarefoundation.github.io/autoware-documentation/main/tutorials/ad-hoc-simulation/rosbag-replay-simulation/).
2. AWSIM

## Evaluation

In AWSIM (Nishishinjuku), the initial pose estimation was performed for each method 200 times, and the final best values of `ndt_align_srv` are obtained.

![compare_histogram_score](https://github.com/autowarefoundation/autoware.universe/assets/28504069/6cf341fa-e0af-4139-827b-6aefeb7682de)

|Method|Mean Score|Process Time|
|:--------:|:---------------:|:----------------:|
|Random|5.938 ± 0.911|1.073 ± 0.043 [sec]|
|TPE|6.233 ± 0.634|1.208 ± 0.045 [sec]|

## Effects on system behavior

Improves the performance of initial position estimation.
Initial position estimation takes only slightly more time.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

KYabuuchi commented:
This is Super Cool :clap:

YamatoAndo commented:
Good Job!

Motsu-san reviewed (commented):
This is a big change. I would like to review this carefully.
Could you let us know the design changes with some functional block diagrams or something?

KYabuuchi commented:
Could you please add tests like the following?  
Since this algorithm is probabilistic, the evaluation criteria will need to be carefully defined.

```cpp
#include "tpe.hpp"

#include <cassert>
#include <iostream>

using Input = TreeStructuredParzenEstimator::Input;
using Trial = TreeStructuredParzenEstimator::Trial;

constexpr double X = 0.1;
constexpr double Y = -0.2;
constexpr double Z = 0.3;

double object_function(const Input & input)
{
  const double dx = (input.x - 0.1);
  const double dy = (input.y + 0.2);
  const double dz = (input.z - 0.3);
  return 6.0 - (dx * dx + dy * dy + dz * dz);
}

int main()
{
  const double stddev = 1;
  TreeStructuredParzenEstimator tpe(10, stddev, stddev, stddev, stddev, stddev);

  for (int i = 0; i < 200; ++i) {
    Input input = tpe.get_next_input();

    Trial trial;
    trial.input = input;
    trial.score = object_function(input);
    tpe.add_trial(trial);

    std::cout << "index: " << i << " score " << trial.score << " x: " << input.x
              << " y: " << input.y << " z: " << input.z << std::endl;
  }

  Input optimal_input = tpe.get_next_input();
  constexpr double EPSILON = 0.1;
  assert(std::abs(optimal_input.x - X) < EPSILON);
  assert(std::abs(optimal_input.y - Y) < EPSILON);
  assert(std::abs(optimal_input.z - Z) < EPSILON);
}

```

KYabuuchi commented on localization/ndt_scan_matcher/include/ndt_scan_matcher/tree_structured_parzen_estimator.hpp:
```diff
@@ -0,0 +1,86 @@
+  {
+    Input input;
+    Score score;
+    int64_t id;
```
`int64_t id;` is not used anywhere. please remove. :pray:

SakodaShintaro replied:
Fixed when changing to generic TPE.

KYabuuchi commented on localization/ndt_scan_matcher/src/tree_structured_parzen_estimator.cpp:
```diff
@@ -0,0 +1,233 @@
+  return result;
+}
+
+double TreeStructuredParzenEstimator::fix_angle(const double angle)
```
Please use [tier4_autoware_utils::normalizeRadian()](https://github.com/autowarefoundation/autoware.universe/blob/e32ed7252a4a8aa4ed171d8714ffa73fdbcab2fc/common/tier4_autoware_utils/include/tier4_autoware_utils/math/normalization.hpp)

kminoda commented on localization/ndt_scan_matcher/src/tree_structured_parzen_estimator.cpp:
```diff
@@ -0,0 +1,233 @@
+  return best_input;
+}
+
+double TreeStructuredParzenEstimator::acquisition_function(const Input & input)
```
Could you elaborate more on this function?
Maybe renaming the function improves a lot, since `acquisition_function` sounds very general. (Not sure if I understood correctly, but something like `get_predicted_score_from_prior_results` for instance.)

SakodaShintaro replied:
The term `acquisition_function` is often used in the context of Bayesian optimization. To be precise, what is implemented here is a function that returns the score of the selection criteria resulting from the transformation of Expectation Improvement (EI), a kind of acquisition functions. So naming it is a difficult task, but I think it is not so bad.

SakodaShintaro replied:
But I renamed to make it better.
https://github.com/autowarefoundation/autoware.universe/pull/5078/commits/8928c88c1df03776dc496b358d15c1fbfb1ee4d6

kminoda commented on localization/ndt_scan_matcher/src/tree_structured_parzen_estimator.cpp:
```diff
@@ -0,0 +1,233 @@
+  std::vector<double> scores;
+  for (int64_t i = 0; i < good_num_; i++) {
+    const double diff = trials_[i].score - MIN_GOOD_SCORE;
+    scores.push_back(std::exp(diff));
```
Apparently there is two kinds of score in this class, one is an input score (=score from NDT), and the internal score (=used for internal validation), which makes it a bit confusing. Please consider using two different variable names.

SakodaShintaro replied:
Changed the score for internal use to explicitly state `log_likelihood_ratio`.
https://github.com/autowarefoundation/autoware.universe/commit/8928c88c1df03776dc496b358d15c1fbfb1ee4d6

kminoda commented on localization/ndt_scan_matcher/src/tree_structured_parzen_estimator.cpp:
```diff
@@ -0,0 +1,233 @@
+  return r;
+}
+
+double TreeStructuredParzenEstimator::log_gaussian_pdf(
+  const Input & input, const Input & mu, const Input & sigma)
```
Please add `const` or make this function independent from TreeStructuredParzenEstimator (fix_angle function as well)
```suggestion
double TreeStructuredParzenEstimator::log_gaussian_pdf(
  const Input & input, const Input & mu, const Input & sigma) const
```

SakodaShintaro replied:
Fixed when changing to generic TPE.

kminoda commented on localization/ndt_scan_matcher/src/tree_structured_parzen_estimator.cpp:
```diff
@@ -0,0 +1,233 @@
+  good_num_ = std::min(good_num_, static_cast<int64_t>(trials_.size() * MAX_GOOD_RATE));
+}
+
+TreeStructuredParzenEstimator::Input TreeStructuredParzenEstimator::get_next_input()
```
Please add const
```suggestion
TreeStructuredParzenEstimator::Input TreeStructuredParzenEstimator::get_next_input() const
```

SakodaShintaro replied:
Fixed in https://github.com/autowarefoundation/autoware.universe/pull/5078/commits/d0590cf656ae14e6c5abf25cb4e2aa109d2aad9d

kminoda commented on localization/ndt_scan_matcher/src/tree_structured_parzen_estimator.cpp:
```diff
@@ -0,0 +1,233 @@
+  return best_input;
+}
+
+double TreeStructuredParzenEstimator::acquisition_function(const Input & input)
```
Please add const

```suggestion
double TreeStructuredParzenEstimator::acquisition_function(const Input & input) const
```

SakodaShintaro replied:
Fixed in https://github.com/autowarefoundation/autoware.universe/pull/5078/commits/d0590cf656ae14e6c5abf25cb4e2aa109d2aad9d

kminoda commented on localization/ndt_scan_matcher/src/tree_structured_parzen_estimator.cpp:
```diff
@@ -0,0 +1,233 @@
+  std::discrete_distribution<int64_t> dist_good(scores.begin(), scores.end());
+
+  Input best_input;
+  double best_score = -1e9;
```
```suggestion
  double best_score = std::numeric_limits<double>::min();
```

SakodaShintaro replied:
Since `log_likelihood_ratio` can be negative, changed to 
`std::numeric_limits<double>::lowest();`.

kminoda commented on localization/ndt_scan_matcher/src/tree_structured_parzen_estimator.cpp:
```diff
@@ -0,0 +1,233 @@
+  good_num_ = std::count_if(trials_.begin(), trials_.end(), [this](const Trial & trial) {
+    return trial.score > MIN_GOOD_SCORE;
+  });
+  good_num_ = std::min(good_num_, static_cast<int64_t>(trials_.size() * MAX_GOOD_RATE));
```
May I ask the rationale behind this cap? I was not sure if this operation is OK given the following part in `acquisition_function`:
```
    const double v = trials_[i].score - MIN_GOOD_SCORE;
    const double w = std::min(std::abs(v), 1.0);
```

SakodaShintaro replied:
Changed to default TPE weighting
https://github.com/autowarefoundation/autoware.universe/blob/8928c88c1df03776dc496b358d15c1fbfb1ee4d6/localization/ndt_scan_matcher/src/tree_structured_parzen_estimator.cpp#L183-L213

kminoda commented on localization/ndt_scan_matcher/config/ndt_scan_matcher.param.yaml:
```diff
@@ -43,6 +43,11 @@
+    # The number of initial random trials in the TPE (Tree-Structured Parzen Estimator).
+    # This value should be equal to or less than 'initial_estimate_particles_num' and more than 0.
+    # If it is equal to 'initial_estimate_particles_num', the search will be the same as a full random search.
+    n_startup_trials: 30
```
Please update readme parameters as well

SakodaShintaro replied:
Fixed in https://github.com/autowarefoundation/autoware.universe/pull/5078/commits/9887fcf41aa1ae2ed75d667dcb8f15df5afa35a9

kminoda commented on localization/ndt_scan_matcher/src/tree_structured_parzen_estimator.cpp:
```diff
@@ -0,0 +1,233 @@
+// Copyright 2023 Autoware Foundation
```
Could you add any reference to TPE so that we can review the internal algorithm as well? (Maybe something like https://papers.nips.cc/paper_files/paper/2011/file/86e8f7ab32cfd12577bc2619bc635690-Paper.pdf?)

kminoda reviewed (commented):
Thanks for awesome job :+1: 

Left some minor comments, but overall looks good to me.

SakodaShintaro commented:
@KYabuuchi 
> Could you please add tests like the following? Since this algorithm is probabilistic, the evaluation criteria will need to be carefully defined.

I think this point is very critical.
The TPE I implemented this time takes advantage of the prior knowledge that a TP of approximately 6 is a good score for the NDT, but that makes it not a normal TPE and makes it difficult to validate with benchmark functions.

Although it is more difficult to use prior knowledge, I have changed my mind that TPE should be implemented as normally as possible and validate with benchmark functions.

Please give me some time to fix the implementation.

SakodaShintaro commented:
The `is_loop_variable` is added in the constructor to inform TPE that yaw is an angle linked between -1 and 1. 
This has resulted in better performance than random.

Minor fix will be made in the future.

SakodaShintaro commented:
@KYabuuchi 

I added test on TPE. https://github.com/autowarefoundation/autoware.universe/pull/5078/commits/d56c845665589a384815060a821122a98cf6d549

You can test as following commands.

```
colcon build --symlink-install --cmake-args -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING=ON --packages-select ndt_scan_matcher
colcon test --packages-select ndt_scan_matcher
colcon test-result --verbose
```

The details of the result are output to directories such as `autoware/log/test_2023-10-05_18-22-43/ndt_scan_matcher/stdout.log`

SakodaShintaro commented:
@Motsu-san 
> This is a big change. I would like to review this carefully. Could you let us know the design changes with some functional block diagrams or something?
I added "Summary of Changes" figure to pull request description.

KYabuuchi commented on localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:
```diff
@@ -784,15 +788,53 @@ geometry_msgs::msg::PoseWithCovarianceStamped NDTScanMatcher::align_using_monte_
-  // generateParticle
-  const auto initial_poses =
-    create_random_pose_array(initial_pose_with_cov, initial_estimate_particles_num_);
+  const double CONVERT_COEFF = std::sqrt(2);
+
+  const auto base_rpy = get_rpy(initial_pose_with_cov);
+  const Eigen::Map<const RowMatrixXd> covariance = {
+    initial_pose_with_cov.pose.covariance.data(), 6, 6};
+  const double stddev_x = CONVERT_COEFF * std::sqrt(covariance(0, 0));
+  const double stddev_y = CONVERT_COEFF * std::sqrt(covariance(1, 1));
+  const double stddev_z = CONVERT_COEFF * std::sqrt(covariance(2, 2));
+  const double stddev_roll = CONVERT_COEFF * std::sqrt(covariance(3, 3));
+  const double stddev_pitch = CONVERT_COEFF * std::sqrt(covariance(4, 4));
+
+  auto uniform_to_normal = [](const double uniform) {
+    assert(-1.0 <= uniform && uniform <= 1.0);
+    constexpr double epsilon = 1.0e-6;
+    const double clamped = std::clamp(uniform, -1.0 + epsilon, 1.0 - epsilon);
+    return boost::math::erf_inv(clamped);
+  };
+
+  auto normal_to_uniform = [](const double normal) { return boost::math::erf(normal); };
```
These conversions are difficult to understand. It is not intuitive why the conversion coefficient is the SQRT(2) and why the conversions between uniform and normal distributions are necessary.

Is it possible to perform these conversions in the `TreeStructuredParzenEstimator` class?
For example, adding modes for normal distribution and uniform distribution to the class.

SakodaShintaro replied:
Added a comment and changed the code slightly.
I am still considering whether to change the structure.
https://github.com/autowarefoundation/autoware.universe/pull/5078/commits/00b0adc3c7935780afa69483f0d734fad458687a

KYabuuchi commented on localization/ndt_scan_matcher/include/ndt_scan_matcher/tree_structured_parzen_estimator.hpp:
```diff
@@ -0,0 +1,80 @@
+  static constexpr double MAX_VALUE = 1.0;
+  static constexpr double MIN_VALUE = -1.0;
+  static constexpr double VALUE_WIDTH = MAX_VALUE - MIN_VALUE;
+  static constexpr int64_t N_EI_CANDIDATES = 100;
```
`EI` is a little bit unclear. Could you add a comment explaining what it stands for?
Is it "Expected Improvement" right?

SakodaShintaro replied:
Yes. The name of this variable is also aligned with optuna.
https://github.com/optuna/optuna/blob/fba23b19ca99702346bec078f3f0bf97eeb07946/optuna/samplers/_tpe/sampler.py#L118

KYabuuchi replied:
I see :ok_hand:.  It is acceptable now.

KYabuuchi commented on localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:
```diff
@@ -806,6 +848,23 @@ geometry_msgs::msg::PoseWithCovarianceStamped NDTScanMatcher::align_using_monte_
+    geometry_msgs::msg::Vector3 rpy = get_rpy(pose);
+
+    TreeStructuredParzenEstimator::Input result(is_loop_variable.size());
+    result[0] = (pose.position.x - initial_pose_with_cov.pose.pose.position.x) / stddev_x;
+    result[1] = (pose.position.y - initial_pose_with_cov.pose.pose.position.y) / stddev_y;
+    result[2] = (pose.position.z - initial_pose_with_cov.pose.pose.position.z) / stddev_z;
+    result[3] = (rpy.x - base_rpy.x) / stddev_roll;
+    result[4] = (rpy.y - base_rpy.y) / stddev_pitch;
+    result[5] = (rpy.z - base_rpy.z) / M_PI;
+    result[0] = normal_to_uniform(result[0]);
+    result[1] = normal_to_uniform(result[1]);
+    result[2] = normal_to_uniform(result[2]);
+    result[3] = normal_to_uniform(result[3]);
+    result[4] = normal_to_uniform(result[4]);
```
Only `yaw(result[5])` is a loop variable, so the conversion method is different, right? 
Please add the reason in a comment.

SakodaShintaro replied:
That is correct. I have added comments and some refactoring to make the intent clearer.
https://github.com/autowarefoundation/autoware.universe/pull/5078/commits/696785099bee16b844c0ec02ead8671a5f6d933c

KYabuuchi replied:
Thank you. It is now easier to understand. :clap:

KYabuuchi commented on localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp:
```diff
@@ -784,15 +788,53 @@ geometry_msgs::msg::PoseWithCovarianceStamped NDTScanMatcher::align_using_monte_
+
+  auto normal_to_uniform = [](const double normal) { return boost::math::erf(normal); };
+
+  // Optimizing (x, y, z, roll, pitch, yaw) 6 dimensions.
+  // The last dimension (yaw) is a loop variable.
+  const std::vector<bool> is_loop_variable = {false, false, false, false, false, true};
```
I think both `roll` and `pitch` are loop variables, too. But since their standard deviations are enough small, there is no problem, right? 
 If so, please describe so in comment.

SakodaShintaro replied:
Exactly right.
Added comment.
https://github.com/autowarefoundation/autoware.universe/pull/5078/commits/e484b9f2eea26dc55f413e84e3465fe8a7d6e64e

KYabuuchi commented on localization/ndt_scan_matcher/include/ndt_scan_matcher/tree_structured_parzen_estimator.hpp:
```diff
@@ -0,0 +1,80 @@
+  static constexpr double MIN_VALUE = -1.0;
+  static constexpr double VALUE_WIDTH = MAX_VALUE - MIN_VALUE;
+  static constexpr int64_t N_EI_CANDIDATES = 100;
+  static constexpr double PRIOR_WEIGHT = 0.0;
```
The `PRIOR_WEIGHT` is always set to 0.0 and has no impact on the estimation now. 
Is this implemented because it might be used in the future, or is it simply forgotten to be removed?

KYabuuchi replied:
It makes sense. Thank you for explaining!

KYabuuchi reviewed (commented):
@SakodaShintaro  This will be my final question. 
Once this is resolved, I will be ready to approve.

KYabuuchi reviewed (approved):
LGTM

kminoda commented:
Just realized that it may be better to change the name of the function `align_using_monte_carlo` 😅
